README.md
# ExYapay

Elixir client that integrates with Yapay.

## Installation

```elixir
def deps do
  [
    {:ex_yapay, "~> 1.0.1"}
  ]
end
```

## Usage

### Configuration

Add the following config to your config.exs file:

```elixir
config :ex_yapay,
  base_url: System.get_env("YAPAY_URL") || "https://intermediador.yapay.com.br",
  request_post_prefix: System.get_env("YAPAY_REQUEST_POST_PREFIX") || "tc.",
  ibrowse_opts: [],
  timeout: 10000
```

Both `YAPAY_URL` and `YAPAY_REQUEST_POST_PREFIX` will need to be set differently depending on the environment:

#### Sandbox

```bash
YAPAY_URL=https://intermediador-sandbox.yapay.com.br
YAPAY_REQUEST_POST_PREFIX=tc-
```

#### Production

```bash
YAPAY_URL=https://intermediador.yapay.com.br
YAPAY_REQUEST_POST_PREFIX=tc.
```

In case none of these are set, the environment you are using is assumed to be production.

### Create a Transaction

The function `create_transaction` expects a map containing an account and a list of products, like so:

```elixir
attributes = %{
  account: %{token_account: "3f3f2f0d347797b"},
  products: [
    %{
      code: 1,
      description: "notebook",
      extra: nil,
      price_unit: 10,
      quantity: 1,
      sku: 123,
      url_img: "https://some.product/url"
    }
  ],
  reseller_token: "a1w0l2l1231lpw0pa",
  shipping_price: 23.39,
  shipping_type: "Correios SEDEX",
  url_notification: "http://localhost:4003/notifications/status?site_id=123&cart_id=445"
}
```

In the struct described above, the mandatory values to start the process are `token_account`, `description`, `price_unit` and `quantity`. Values other than those can be `nil`, but their keys are required as well.

In order to receive the checkout url, the only call to be made is:

```elixir
ExYapay.create_transaction(attributes)
```

Which will return either:

```elixir
{:ok, "some.checkout/url"}
```

Or

```elixir
{:error, %{body: 400, body: "message the reflects what went wrong"}}
```

### Get a Transaction

The function `get_transaction` expects the `token_account` and `token_transaction`:

```elixir
ExYapay.get_transaction("6f43694d9ec6057", "9342ef911dd843e7a2fae4a41357727f")
{:ok,
 %ExYapay.Resources.Transaction{
   customer: %{
     addresses: [
       %{
         city: "Campo Grande",
         completion: "",
         neighborhood: "Centro",
         number: "2434",
         postal_code: "79002003",
         state: "MS",
         street: "Avenida Calógeras"
       }
     ],
     cnpj: "",
     company_name: "",
     contacts: [%{type_contact: "W", value: "1135440444"}],
     cpf: "11122233388",
     email: "your.email@example.com",
     name: "Locaweb",
     trade_name: "Locaweb"
   },
   free: "|POST| |checkout|",
   order_number: "1576009266",
   payment: %{
     date_approval: 1576009320,
     date_payment: 1576009320,
     linha_digitavel: nil,
     payment_method_id: 4,
     payment_method_name: "Mastercard",
     payment_response: "Mensagem de venda fake",
     price_original: "208.0",
     price_payment: "216.33",
     split: 3,
     tid: "1233",
     url_payment: nil
   },
   refunds: [],
   status_id: 6,
   status_name: "Aprovada",
   token_transaction: "9342ef911dd843e7a2fae4a41357727f",
   transaction_id: 313439
 }}
```

If an error occurs, the answer will be as follows:

```elixir
ExYapay.get_transaction("6f43694d9ec6057", "123")
{:error, %{body: "Transação não encontrada", status: 404}}
```

## Running tests

```bash
mix test
```

## Running code formatter

```bash
mix format
```

## Credo

Credo is a static code analysis tool for the Elixir language, to run credo:

```
$ mix credo --strict
```

## Contributing

Check out the [Contributing](CONTRIBUTING.md) guide.

## License

ExYapay is released under the MIT license. See the [License](LICENSE.md) file.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [1.0.1] - 2020-08-20
### Changed
- Facilitate switching between sandbox and production environments changing `request_post_prefix` and `request_get_prefix` from module attributes to private functions.

## [1.0.0] - 2020-05-18
### Added
- Add `travis CI`.
- Publish lib to Hex.pm.
- Add MIT license to the project.
- Add `Contributing` section to the `Readme.md`.

### Changed
- Rename Yapay context and application name to `ExYapay`.
- Update `README.md`.

## [0.4.2] - 2020-05-18
### Changed
- Add `locawebcommon` info to the project.

## [0.4.1] - 2020-03-04
### Fixed
- Fix character issue in the yapay transaction creation.

### Changed
- Update elixir version to [1.10](https://elixir-lang.org/blog/2020/01/27/elixir-v1-10-0-released/).

## [0.4.0] - 2019-12-23
### Added
- Add `Transaction.get(token_account, token_transaction)` to get a transaction.

### Removed
- Remove `Transaction.create` with the `Yapay` struct.

## [0.3.0] - 2019-12-03
### Added
- Add support for the attributes: `shipping_price` and `shipping_type`.

## [0.2.0] - 2019-11-18
### Added
- Add support for the attributes: `reseller_token` and `url_notification`.
- Add support for `ibrowse` options and `timeout`.

### Fixed
- Request error handling.

### Changed
- Change function `Yapay.create_transaction/2` to support: `Yapay struct` or `map`.

## [0.1.0] - 2019-08-26
### Added
- Integration with Yapay's transaction API.

## [0.0.1] - 2019-08-13
### Added
- Bootstrap application.

[Unreleased]: https://github.com/locaweb/ex-yapay/compare/master...HEAD
[1.0.1]: https://github.com/locaweb/ex-yapay/compare/v1.0.0..v1.0.1
[1.0.0]: https://github.com/locaweb/ex-yapay/releases/tag/v1.0.0


lib/ex_yapay.ex
defmodule ExYapay do
  @moduledoc """
  A client for the `Yapay Intermediador API`.

  More information can be found on the yapay doc: http://dev.yapay.com.br/intermediador/apis.
  """

  alias ExYapay.Resources
  alias ExYapay.Resources.{Account, Product}
  alias ExYapay.Transaction

  @typep options :: [parser: atom(), request: atom()]
  @typep reason :: %{body: String.t(), status: integer()}

  @type create_attributes :: %{
          account: %Account{},
          products: [%Product{}],
          reseller_token: String.t(),
          shipping_price: float(),
          shipping_type: String.t(),
          url_notification: String.t()
        }

  @doc """
  Creates a transaction via POST request to Yapay and return its checkout url.

  ## Examples

    iex> attributes = %{
      account: %{token_account: "3f3f2f0d347797b"},
      products: [
        %{
          code: 1,
          description: "notebook",
          extra: nil,
          price_unit: 10,
          quantity: 1,
          sku: 123,
          url_img: "some.product/url"
        }
      ],
      reseller_token: "a1w0l2l1231lpw0pa",
      shipping_price: 23.39,
      shipping_type: "Correios SEDEX",
      url_notification: "http://localhost:4003/notifications/status?site_id=123&cart_id=445"
    }
    iex> ExYapay.create_transaction(attributes)
    {:ok, "https://checkout.yapay.com.br/transacao/ta679ae0f5f25e814a0d79bd35ee292d4"}

  """
  @spec create_transaction(create_attributes(), options()) ::
          {:ok, String.t()} | {:error, reason()}
  defdelegate create_transaction(attributes, opts \\ []), to: Transaction, as: :create

  @doc """
  Gets a transaction by `token_account` and `token_transaction`.

  ## Examples

    iex> ExYapay.get_transaction("6f43694d9ec6057", "9342ef911dd843e7a2fae4a41357727f")
    {:ok,
     %ExYapay.Resources.Transaction{
       customer: %{
         addresses: [
           %{
             city: "Campo Grande",
             completion: "",
             neighborhood: "Centro",
             number: "2434",
             postal_code: "79002003",
             state: "MS",
             street: "Avenida Calógeras"
           }
         ],
         cnpj: "",
         company_name: "",
         contacts: [%{type_contact: "W", value: "1135440444"}],
         cpf: "11122233388",
         email: "example@example.com.br",
         name: "Criador de Sites",
         trade_name: "Criador de Sites"
       },
       free: "|POST| |checkout|",
       order_number: "1576009266",
       payment: %{
         date_approval: 1576009320,
         date_payment: 1576009320,
         linha_digitavel: nil,
         payment_method_id: 4,
         payment_method_name: "Mastercard",
         payment_response: "Mensagem de venda fake",
         price_original: "208.0",
         price_payment: "216.33",
         split: 3,
         tid: "1233",
         url_payment: nil
       },
       refunds: [],
       status_id: 6,
       status_name: "Aprovada",
       token_transaction: "9342ef911dd843e7a2fae4a41357727f",
       transaction_id: 313439
     }}

    iex> ExYapay.get_transaction("6f43694d9ec6057", "123")
    {:error, %{body: "Transação não encontrada", status: 404}}

  """
  @spec get_transaction(String.t(), String.t(), options()) ::
          {:ok, Resources.Transaction.t()} | {:error, reason()}
  defdelegate get_transaction(token_account, token_transaction, opts \\ []),
    to: Transaction,
    as: :get
end
